src/testrender/raytracer.h

```diff
@@ -278,6 +278,24 @@ struct Scene {
         return ((1 - u - v) * na + u * nb + v * nc).normalize();
     }
 
+    // Project differentials onto surface defined by N
+    OSL_HOSTDEVICE
+    void project(Dual2<Vec3>& p, const Vec3& N, const Vec3& I) const
+    {
+        float cosx = dot(p.dx().normalized(), N);
+        float cosy = dot(p.dy().normalized(), N);
+        float cosI = dot(I.normalized(), N);
+
+        if (abs(cosI)>1e-3 )
```

Comment:
The overload specification of `std::abs` is a bit convoluted in C++ - it appears to be working here but I prefer to use `fabsf()` to be explicit that we want the float to float version.

Constant should be `1e-3f` to avoid extra casts to double precision.